include/storage/fbw.h

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright (c) 2017, 2020 Nordic Semiconductor ASA
+ * Copyright (c) 2017 Linaro Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief Public API for flash buffered writes
+ */
+
+#ifndef ZEPHYR_INCLUDE_STORAGE_FBW_H_
+#define ZEPHYR_INCLUDE_STORAGE_FBW_H_
+
+/**
+ * @brief Abstraction over flash buffered writes.
+ *
+ * @defgroup fbw flash buffered write Interface
+ * @{
+ */
+
+#include <stdbool.h>
+#include <drivers/flash.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef int (*fbw_callback_t)(u8_t *buf, size_t len, size_t offset);
+
+/**
+ * @brief Structure for flash buffered writes
+ *
+ * Users should treat these structures as opaque values and only interact
+ * with them through the below API.
+ */
+struct fbw_ctx {
+	u8_t *buf; /* Write buffer */
+	size_t buf_len; /* Length of write buffer */
+	size_t buf_bytes; /* Number of bytes currently stored in write buf */
+	struct device *fdev; /* Flash device */
+	size_t bytes_written; /* Number of bytes written to flash */
+	size_t offset; /* Offset from base of flash device to write area */
+	size_t available; /* Available bytes in write area */
+	fbw_callback_t callback; /* Callback invoked on completed write op */
+#ifdef CONFIG_FBW_ERASE
+	off_t off_last; /* Last erased offset */
+#endif
+};
+
+/**
+ * @brief Initialize context needed for buffered write to flash.
+ *
+ * @param ctx context to be initialized
+ * @param fdev Flash device to operate on
+ * @param buf Write buffer
+ * @param buf_len Length of write buffer
+ * @param offset Offset within flash device
+ * @param size Size of region within flash device
+ * @param cb Callback to be invoked on completed write operations
+ *
+ * @return  0 on success, negative errno code on fail
+ */
+int fbw_init(struct fbw_ctx *ctx, struct device *fdev, u8_t *buf,
+	     size_t buf_len, size_t offset, size_t size, fbw_callback_t cb);
+/**
+ * @brief Read number of bytes written to the flash.
+ *
+ * @param ctx context
+ *
+ * @return Number of bytes written to flash.
+ */
+size_t fbw_bytes_written(struct fbw_ctx *ctx);
+
+/**
+ * @brief  Process input buffers to be written to flash device in single blocks.
+ * Will store remainder between calls.
+ *
+ * A final call to this function with flush set to true
+ * will write out the remaining block buffer to flash.
+ *
+ * @param ctx context
+ * @param data data to write
+ * @param len Number of bytes to write
+ * @param flush when true this forces any buffered data to be written to flash
+ *
+ * @return  0 on success, negative errno code on fail
+ */
+int fbw_write(struct fbw_ctx *ctx, u8_t *data, size_t len, bool flush);
+
+/**
+ * Erase the flash area progressively
+ *
+ * This function erases a flash page to which an offset belongs if this page
+ * wasn't erased before by this module.
```

Comment:
This part I find a bit confusing when trying to imagine what this API does from only these doxygen comments. Some questions that come into my mind when looking at this API with a blank slate:

- Does that mean I have to call fbw_init() again to allow an erase of the same sector? Will this return an error if I try to erase something again, or is behavior undefined?
- Does this API somehow remember which sectors it has erased? If so, where does the memory for that get allocated from? If not, does the context just remember a most recent sector written to or something?

Reply:
I tried to rephrase the docs. Note that the only reason this is exposed is because this is the exact functionality required by mcuboot flash image writers, and I prefer providing a "weird" function rather than duplicating code.